buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/apiinfo/ApiInformationTask.java

```diff
@@ -0,0 +1,93 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.gradle.plugins.apiinfo;
+
+import org.gradle.api.DefaultTask;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.InputFile;
+import org.gradle.api.tasks.OutputDirectory;
+import org.gradle.api.tasks.OutputFile;
+import org.gradle.api.tasks.TaskAction;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.Arrays;
+
+/**
+    Task generates the api diff of the current source code against the api.txt file stored
+    alongside the project's src directory.
+ */
+public abstract class ApiInformationTask extends DefaultTask {
+
+    @Input
+    abstract String getMetalavaBinaryPath();
+
+    @InputFile
+    abstract File getApiTxt();
+
+    @Input
+    abstract String getSourcePath();
+
+    @InputFile
+    abstract File getBaselineFile();
+
+    @Input
+    abstract boolean getUpdateBaseline();
+
+    @OutputFile
+    abstract File getOutputFile();
+
+    abstract void setSourcePath(String value);
+
+    abstract void setBaselineFile(File value);
+
+    abstract void setUpdateBaseline(boolean value);
+
+    abstract void setMetalavaBinaryPath(String value);
+
+    abstract void setApiTxt(File value);
+
+    abstract void setOutputFile(File value);
+
+
+    @TaskAction
+    void execute() {
+        File outputFileDir = getOutputFile().getParentFile();
+        if(!outputFileDir.exists()) {
+            outputFileDir.mkdirs();
+        }
+        String cmdTemplate = "%s --source-path %s --check-compatibility:api:current %s --format=v2 --baseline %s --no-color";
+        if(getUpdateBaseline()) {
+            cmdTemplate = "%s --source-path %s --check-compatibility:api:current %s --format=v2 --update-baseline %s --no-color";
```

Comment:
IMO, there is a lot of boilerplate here that comes down to 2 commands and their respective arguments. 
```
%s --source-path %s --api %s --format=v2 --baseline %s
```
```
 "%s --source-path %s --check-compatibility:api:current %s --format=v2 --update-baseline %s --no-color";
```
I think we can remove some cruft by modelling this as 
```java
// This task executes in 2 modes.
// 1. Generate api text mode
// 2. Compare api text mode 
public abstract class MetalavaExecutionTask extends DefaultTask {
    @Input
    abstract String getMetalavaBinaryPath();

    @InputFile
    abstract File getApiTxt();

    @Input
    abstract String getSourcePath();

    @InputFile
    abstract File getBaselineFile();

    @Input
    abstract boolean getUpdateBaseline();

    @Input
    abstract enum getMode();


    @OutputFile
    abstract File getOutputFile();

    abstract void setSourcePath(String value);

    abstract void setBaselineFile(File value);

    abstract void setUpdateBaseline(boolean value);

    abstract void setMetalavaBinaryPath(String value);

    abstract void setApiTxt(File value);

    abstract void setOutputFile(File value);

    abstract void setMode(Mode mode);

    private static enum Mode {
        GENERATE, COMPARE
    }

    @TaskAction
    void execute() {
        File outputFileDir = getOutputFile().getParentFile();
        if(!outputFileDir.exists()) {
            outputFileDir.mkdirs();
        }
        
        String cmd = mode == Mode.GENERATE ? "--check-compatibility:api:current %s" : "--api %s"
        String baseline = "--baseline %s" : "--update-baseline %s"
        String format = "--format=v2"
        String sourcePath = "--source-path %s"

        String cmdTemplate = ["%s", sourcePath, cmd, baselineArg, "--no-color"].join(" ")

        String cmdToRun = String.format(cmdTemplate, getMetalavaBinaryPath(), getSourcePath(), getApiTxt().getAbsolutePath(), getBaselineFile().getAbsolutePath());

        getProject().exec(spec-> {
            spec.setCommandLine(Arrays.asList(cmdToRun.split(" ")));
            try {
                spec.setStandardOutput(new FileOutputStream(getOutputFile()));
            } catch (FileNotFoundException e) {
                getLogger().error(e.toString());
            }
        });


        
    }

}
```

Reply:
in the ci command we would need to run the task to getApiInformation
and then run the task to generateApiInformation in this order
so we would need two tasks. However we can merge the two task classes into one though

Reply:
And there is no purpose of outputFile in the generateApiInformation Task

Reply:
The structure of the second task has changed now. It involves producing the api.txt file and then comparing the already existing api.txt with that. I feel now the structure of the two tasks are fairly different